StringRep.hs

```diff
@@ -13,6 +14,76 @@ import qualified Miso.String as MS
 import qualified Data.Text.Lazy.Builder as B
 import qualified Data.Text.Lazy as L
 import Terms
+import qualified Text.Earley as EP
+import qualified Text.Earley.Mixfix as EPM
+
+data Token = LParen | RParen | Word MS.MisoString | Dot | Binder MS.MisoString deriving (Show, Eq)
+data TermH = WordH (EPM.Holey Token) | BinderH Token | LamH TermH TermH | AppH TermH [TermH] deriving (Show, Eq)
+
+holey :: String -> EPM.Holey MS.MisoString
+holey ""       = []
+holey ('_':xs) = Nothing : holey xs
+holey xs       = Just (MS.toMisoString i) : holey rest
+  where (i, rest) = span (/= '_') xs
+
+mixfixTable :: [[(EPM.Holey MS.MisoString, EPM.Associativity)]]
+mixfixTable = (map . map) (first holey)
+  [ [("_->_",          EPM.RightAssoc)]
+  , [("_,_",           EPM.NonAssoc)]
+  , [("if_then_else_", EPM.RightAssoc)]
+  , [("_|-_:_",        EPM.NonAssoc)]
+  , [("_+_",           EPM.LeftAssoc)]
+  , [("_*_",           EPM.LeftAssoc)]
+  ]
+
+grammar :: EP.Grammar r (EP.Prod r Token Token TermH)
+grammar = mdo
+  identB    <- EP.rule $ BinderH <$> EP.satisfy isLegalBinder
```

Comment:
It looks to me like you don't need `BinderH`. Instead, you could just have `identB` always return a string and make `LamH` take a string and a term, so the only thing you have to do in `postProc` is handle `deBruijn` indices and fold the `Ap`

Reply:
It may be possible to just parse `Term` directly but always parse words as `Const`, then in a separate pass go through with the context and add deBruijn indices as necessary.

Reply:
Resolved! I learnt `foldl'` but never thought it could be used like that. Thanks!